apps/passport-server/src/services/generic-issuance/pipelines/PODPipeline/PODPipeline.ts

```diff
@@ -0,0 +1,439 @@
+    const atoms = await this.db.load(this.id);
+    const matchingAtoms: PODAtom[] = [];
+
+    for (const atom of atoms) {
+      if (atom.matchTo.type === "none") {
+        // No filter, so all atoms match to all credentials
+        matchingAtoms.push(atom);
+      } else if (atom.matchTo.type === "email") {
+        if (
+          atom.entries[atom.matchTo.entry].value.toString().toLowerCase() ===
+          email.toLowerCase()
+        ) {
+          matchingAtoms.push(atom);
+        }
+      } else if (atom.matchTo.type === "semaphoreID") {
+        if (
+          atom.entries[atom.matchTo.entry].value.toString() ===
+          semaphoreId.toString()
+        ) {
+          matchingAtoms.push(atom);
+        }
+      }
+    }
+
+    return matchingAtoms;
+  }
+
+  /**
+   * Get a cached PCD for the given ID. IDs are derived from the POD
+   * entries and pipeline ID. We also want to ensure that the POD was signed
+   * with the current EdDSA private key, so we include that in the cache key.
+   *
+   * @param id The id of the PCD to get
+   * @returns The cached PCD, or undefined if it is not cached
+   */
+  private async getCachedPCD(id: string): Promise<SerializedPCD | undefined> {
+    const key = await getHash(`${id}${this.eddsaPrivateKey}`);
+    const cacheEntry = await this.cacheService.getValue(key);
+    if (cacheEntry) {
+      return JSON.parse(cacheEntry.cache_value);
+    }
+    return undefined;
+  }
+
+  /**
+   * Store a serialized PCD in the cache.
+   *
+   * @param id The id of the PCD to store
+   * @param serializedPCD The PCD to store
+   */
+  private async storeCachedPCD(
+    id: string,
+    serializedPCD: SerializedPCD
+  ): Promise<void> {
+    const key = await getHash(`${id}${this.eddsaPrivateKey}`);
+    await this.cacheService.setValue(key, JSON.stringify(serializedPCD));
+  }
+
+  /**
+   * Return a response to the feed issuance request. This takes the form of
+   * an array of PCD actions that will be applied to the user's PCD collection.
+   *
+   * @param req The request for feed issuance
+   * @returns The response to the feed issuance request
+   */
+  private async feedIssue(
+    req: PollFeedRequest
+  ): Promise<PollFeedResponseValue> {
+    return traced(LOG_NAME, "feedIssue", async (span) => {
+      if (!req.pcd) {
+        throw new Error("missing credential pcd");
+      }
+
+      if (this.definition.options.paused) {
+        return { actions: [] };
+      }
+
+      const credential =
+        await this.credentialSubservice.verifyAndExpectZupassEmail(req.pcd);
+      const { email, semaphoreId } = credential;
+
+      span?.setAttribute("email", email);
+      span?.setAttribute("semaphore_id", semaphoreId);
+
+      // Consumer is validated, so save them in the consumer list
+      const didUpdate = await this.consumerDB.save(
+        this.id,
+        email,
+        semaphoreId,
+        new Date()
+      );
+      if (didUpdate) {
```

Comment:
This is doing the same thing twice?  I don't understand why.

Reply:
This is an error. Removed!